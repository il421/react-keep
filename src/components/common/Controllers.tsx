import React from "react";
import { AlignItems, JustifyContent } from "../../common";
import { FlexBox, IconButton } from "../ui-components";
import "../../styles/components/common/_controllers.scss";
import { NoteType } from "../notes";

interface Controllers extends Partial<Pick<HTMLElement, "className">> {
  isMobile?: boolean;
  openDialog: (type: NoteType) => void;
}

export const Controllers: React.FunctionComponent<Controllers> = ({
  className,
  isMobile = false,
  openDialog,
}) => {
  const handleOnClick = (type: NoteType) => () => openDialog(type);

  return (
    <div
      className={`controllers  ${
        isMobile && "controllers--mobile"
      } ${className}`}
    >
      <FlexBox
        vertical={isMobile}
        justifyContent={JustifyContent.spaceBetween}
        alignItems={AlignItems.center}
      >
        <IconButton
          className="controllers__text"
          onButtonClick={handleOnClick(NoteType.text)}
          icon="align-left"
          size="2x"
        />

        <IconButton
          className="controllers__list"
          onButtonClick={handleOnClick(NoteType.list)}
          icon="check-square"
          size="2x"
        />

        <IconButton
          className="controllers__image"
          onButtonClick={handleOnClick(NoteType.image)}
          icon="image"
          size="2x"
        />
      </FlexBox>
    </div>
  );
};
