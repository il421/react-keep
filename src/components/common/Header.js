import React from 'react';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Search from './Search';

export const Header = ({ startLogout, auth, showSidebar }) => {

  return (
    <header className="header" onClick={(evt) => evt.stopPropagation()}>
      <div className="content-container">
        <div className="header__content">
          <div className="header__sidebar">

            <button className="button button--sidenav" onClick={() => showSidebar()}>
              <FontAwesomeIcon icon="bars" size="2x" />
            </button>

          </div>
          <div>
            <h1 className="header__title">Keep Me</h1>
          </div>
          <div className="header__search">
            <Search />
          </div>
          <div className="header__user-box user-box">
            <div className="user-box__photo">
              <img src={ auth.url } width={ 50 } height={ 50 } />
            </div>
            <div className="user-box__name show-for-mobile">{ auth.name }</div>
            <button className="user-box__logout button button--sidenav" onClick={ startLogout }>
              <FontAwesomeIcon icon="sign-out-alt" size="lg" />
            </button>
          </div>
        </div>
      </div>
    </header>
  );
};


const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    url: state.url,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
