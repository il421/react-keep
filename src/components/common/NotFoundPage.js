import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
  <div className="not-found">
    <div className="not-found__wrapper">
      <div className="not-found__text">
        Sorry, such a page does not exist in the App. <br /><br />
        Try to find something interesting on the Dashboard page.
      </div>
      <Link to="/" className="not-found__link button">Keep me ...</Link>
    </div>
  </div>
);

export default NotFoundPage;
