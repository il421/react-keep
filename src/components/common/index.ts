export * from "./Controllers";
export * from "./DashboardPage";
export * from "./Header";
export * from "./NotFoundPage";
export * from "./Search";
export * from "./SideBar";
