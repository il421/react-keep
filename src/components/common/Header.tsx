import React from "react";
import { connect } from "react-redux";
import { useHistory } from "react-router-dom";
import { stringify } from "query-string";
import { AuthStoreState, Store } from "../../store/store.types";
import { QueryKeys, RouteActions } from "../../routers/Routing";
import { startLogout } from "../../actions/auth";
import Search from "./Search";
import { Controllers } from "./Controllers";
import {
  IconButton,
  UserPhoto,
  ContentContainer,
  LinkButton
} from "../ui-components";
import "../../styles/components/header/_header.scss";
import "../../styles/components/header/_user-box.scss";
import "../../styles/components/common/_visibility.scss";
import { ThunkDispatch } from "redux-thunk";

interface HeaderProps {
  showSidebar: (value: boolean) => void;
}

interface StateProps {
  auth: AuthStoreState;
}

interface DispatchProps {
  startLogout: () => void;
}

type Props = HeaderProps & DispatchProps & StateProps;

const Header: React.FunctionComponent<Props> = ({
  startLogout,
  auth,
  showSidebar
}) => {
  const history = useHistory();

  const onClickHandler = (key: keyof typeof QueryKeys) => {
    const query = stringify({
      [key]: RouteActions.edit
    });
    history.push(`${history.location.pathname}?${query}`);
  };

  return (
    <header className="header" onClick={evt => evt.stopPropagation()}>
      <ContentContainer>
        <div className="header__content">
          <IconButton
            className="header__bar"
            icon="bars"
            size="2x"
            onClick={() => showSidebar(true)}
          />

          <h1 className="header__title">Keep Me</h1>

          <Search wrapperClass="header__search" />

          <Controllers className="hide-for-mobile" />

          <div className="header__user-box user-box">
            {auth.name && auth.url && (
              <UserPhoto
                className="user-box__photo"
                src={auth.url}
                height={50}
                width={50}
              />
            )}
            <LinkButton
              text={auth.name ?? "Update profile"}
              type="button"
              className="user-box__link"
              onClick={() => onClickHandler(QueryKeys.user)}
            />
            <IconButton
              className="user-box__logout"
              icon="sign-out-alt"
              size="lg"
              onClick={startLogout}
            />
          </div>
        </div>
      </ContentContainer>
    </header>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state: Store): StateProps => {
  return {
    auth: state.auth
  };
};

export default connect<StateProps, DispatchProps, HeaderProps, Store>(
  mapStateToProps,
  mapDispatchToProps
)(Header);
