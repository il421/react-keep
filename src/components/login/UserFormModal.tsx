import React from "react";
import { connect } from "react-redux";
import Modal from "react-modal";
import { AlignItems, JustifyContent, nameOf, Placeholders } from "../../common";
import {
  AuthStoreState,
  ModalsStoreState,
  Store,
  UpdateUser,
} from "../../store/store.types";
import { updateUserData } from "../../actions/auth";
import {
  ContentContainer,
  ConfirmButton,
  FlexBox,
  IconButton,
} from "../ui-components";
import { BaseForm, TextInputField, FileFormField } from "../form";
import "../../styles/components/login/_user-form.scss";
import { ThunkDispatch } from "redux-thunk";
import { Field, FormSpy } from "react-final-form";
import { FormApi } from "final-form";
import { toggle } from "../../actions/modals";

interface UserFormValues {
  firstName: string;
  lastName: string;
  photoUrl: string | null;
  uploadingPhoto?: File;
}

interface StateProps {
  auth: AuthStoreState;
  modals: ModalsStoreState;
}

interface FormSpyProps {
  form: FormApi;
  values: UserFormValues;
  initialValues: UserFormValues;
}

interface DispatchProps {
  updateUserData: (data: UpdateUser) => void;
  toggle: (isOpen: boolean) => void;
}

interface UserFormModalProps {}

export type Props = StateProps & UserFormModalProps & DispatchProps;

export const UserFormModal: React.FunctionComponent<Props> = ({
  updateUserData,
  toggle,
  modals,
  auth,
}) => {
  const nameOfField = nameOf<UserFormValues>();
  const NO_AVATAR_URL: string = "img/no_avatar.png";
  const FIELD_ID: string = "avatar";

  const getFormValues = (): UserFormValues => {
    let firstName: string = "";
    let lastName: string = "";

    if (auth.name !== null) {
      firstName = auth.name.split(" ")[0];
      lastName = auth.name
        .split(" ")
        .filter((i) => i !== firstName)
        .join(" ");
    }

    return {
      firstName,
      lastName,
      photoUrl: auth.url,
    };
  };

  const onToggleAction = (isOpen: boolean) => () => toggle(isOpen);

  const getButtons = (isDisable: boolean) => {
    return (
      <>
        <ConfirmButton
          text="Update"
          loading={auth.loading}
          disabled={isDisable}
          className="login-button user-form__button"
        />

        <ConfirmButton
          text="Close"
          type="button"
          onClick={onToggleAction(false)}
          className="login-button user-form__button"
        />
      </>
    );
  };

  const onSubmit = (values: UserFormValues) => {
    const displayName = `${values.firstName} ${values.lastName}`.trim();
    updateUserData({
      displayName: !!displayName ? displayName : null,
      photoFile: values.uploadingPhoto,
      photoURL: values.photoUrl,
      tenantId: auth.uid,
    });
  };

  const handleUploadClick = () => {
    const fileInput = document.getElementById(FIELD_ID);
    fileInput && fileInput.click();
  };

  return (
    <Modal
      isOpen={modals.user.isOpen}
      onRequestClose={onToggleAction(false)}
      className="user-modal"
      ariaHideApp={false}
    >
      <ContentContainer className="user-modal__container">
        <BaseForm<UserFormValues>
          formClassName="login-box__form user-form"
          initialValues={getFormValues()}
          onSubmit={onSubmit}
          getFormActions={getButtons}
        >
          <FlexBox
            justifyContent={JustifyContent.start}
            alignItems={AlignItems.center}
            className="user-form__avatar avatar"
          >
            <FormSpy>
              {({ form, values, initialValues }: FormSpyProps) => {
                return (
                  <>
                    <Field name={nameOfField("photoUrl")}>
                      {() => (
                        <img
                          className="avatar__img"
                          src={
                            values.photoUrl && !values.uploadingPhoto
                              ? values.photoUrl
                              : values.uploadingPhoto
                              ? URL.createObjectURL(values.uploadingPhoto)
                              : NO_AVATAR_URL
                          }
                          width="95"
                          height="95"
                          alt="User photo"
                        />
                      )}
                    </Field>
                    <FlexBox
                      justifyContent={JustifyContent.spaceBetween}
                      alignItems={AlignItems.center}
                      vertical
                    >
                      <IconButton
                        className="avatar__btn"
                        icon="upload"
                        onButtonClick={() => {
                          handleUploadClick();
                        }}
                      />
                      <IconButton
                        className="avatar__btn"
                        icon="times"
                        onButtonClick={() => {
                          form.change(nameOfField("uploadingPhoto"), undefined);
                        }}
                      />
                      {initialValues.photoUrl && (
                        <IconButton
                          id="test-delete-avatar-button"
                          className="avatar__btn"
                          icon="trash"
                          onButtonClick={() => {
                            form.change(nameOfField("photoUrl"), null);
                            form.change(
                              nameOfField("uploadingPhoto"),
                              undefined
                            );
                          }}
                        />
                      )}
                      <FileFormField
                        id={FIELD_ID}
                        name={nameOfField("uploadingPhoto")}
                        className="avatar__field"
                      />
                    </FlexBox>
                  </>
                );
              }}
            </FormSpy>
          </FlexBox>

          <FlexBox
            vertical
            justifyContent={JustifyContent.spaceBetween}
            className="user-form__fields"
          >
            <TextInputField
              name={nameOfField("firstName")}
              type="text"
              placeholder={Placeholders.firstName}
              className="user-form__field"
            />

            <TextInputField
              name={nameOfField("lastName")}
              type="text"
              placeholder={Placeholders.lastName}
              className="user-form__field"
            />
          </FlexBox>
        </BaseForm>
      </ContentContainer>
    </Modal>
  );
};

const mapStateToProps = (state: Store): StateProps => {
  return {
    auth: state.auth,
    modals: state.modals,
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  updateUserData: (data: UpdateUser) => dispatch(updateUserData(data)),
  toggle: (isOpen: boolean) => dispatch(toggle("user", isOpen)),
});

export default connect<StateProps, DispatchProps, UserFormModalProps, Store>(
  mapStateToProps,
  mapDispatchToProps
)(UserFormModal);
