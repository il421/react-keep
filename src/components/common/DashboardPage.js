import React, { useState } from 'react';

import NoteSelection from '../note/NoteSelection';
import NotesList from '../note/NotesList';
import Header from './Header';
import SideBar from './SideBar';


const DashboardPage = () => {
  let [showBar, toggleSidebarDisplay] = useState(false);

  const showSidebar = () => {
    toggleSidebarDisplay(showBar = true);
  };

  const hideSidebar = () => {
    toggleSidebarDisplay(showBar = false);
  };

  return (
    <div className="dashboard">
      <Header showSidebar={ showSidebar } />
      <NoteSelection className="content-container"/>
      <NotesList/>

      <SideBar showBar={ showBar } hideSidebar={ hideSidebar } />
    </div>
  );
};
export default DashboardPage;
