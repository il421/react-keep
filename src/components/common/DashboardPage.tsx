import React from "react";
import { ToastContainer } from "react-toastify";
import { History, Path } from "history";
import { JustifyContent } from "../../common";
import { Header } from "./Header";
import { SideBar } from "./SideBar";
import { Controllers } from "./Controllers";
import UserFormModal from "../login/UserFormModal";
import "../../styles/components/common/_dashboard.scss";
import "../../styles/components/notes/_note-modal.scss";
import { QueryKeys } from "../../routers/Routing";
import { stringify } from "query-string";
import {
  NoteType,
  ImageNoteFormModal,
  TextNoteFormModal,
  ListNoteFormModal,
} from "../notes";
import { FlexBox } from "../ui-components";
import { NotesList } from "../notes/NotesList";

export interface DashboardPageProps {
  history: History;
}

export const onNoteSelected = (options: {
  type: NoteType;
  id: string;
  pathname: string;
  push: (path: Path) => void;
}) => {
  let key: keyof typeof QueryKeys;
  const { type, id, pathname, push } = options;
  switch (type) {
    case NoteType.text:
      key = QueryKeys.text;
      break;
    case NoteType.list:
      key = QueryKeys.list;
      break;
    default:
      key = QueryKeys.image;
  }

  const query = stringify({
    [key]: id,
  });
  push(`${pathname}?${query}`);
};

export const DashboardPage: React.FunctionComponent<DashboardPageProps> = ({
  history,
}): JSX.Element => {
  return (
    <FlexBox
      vertical={true}
      justifyContent={JustifyContent.spaceBetween}
      className="dashboard"
    >
      <Header history={history} />
      <NotesList
        onNoteSelected={(type: NoteType, id: string) =>
          onNoteSelected({
            type,
            id,
            pathname: history.location.pathname,
            push: history.push,
          })
        }
      />

      <SideBar />
      <TextNoteFormModal history={history} />
      <ListNoteFormModal history={history} />
      <ImageNoteFormModal history={history} />
      <UserFormModal />
      <Controllers
        isMobile
        openDialog={(query) =>
          history.push(`${history.location.pathname}?${query}`)
        }
      />

      <ToastContainer autoClose={2000} />
    </FlexBox>
  );
};
