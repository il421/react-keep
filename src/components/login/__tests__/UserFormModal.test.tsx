import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { Props, UserFormModal } from "../UserFormModal";
import { defaultAuthState } from "../../../actions/__tests__/auth.test";
import { triggerInputChange } from "../../../common/testUtils";
import { user } from "../../../testData/users";
import fs from "fs";

let wrapper: ReactWrapper | undefined, props: Props;
const file = fs.readFileSync("./src/testData/test_img.jpg");

beforeEach(() => {
  props = {
    updateUserData: jest.fn(),
    toggleUserModal: jest.fn(),
    auth: { ...defaultAuthState.auth },
  };
  wrapper = mount(<UserFormModal {...props} />);
});

afterEach(() => {
  if (wrapper) {
    wrapper.unmount();
    wrapper = undefined;
  }
});

describe("Rendering", () => {
  test("should render LoginPage correctly", () => {
    if (wrapper) {
      expect(wrapper.debug()).toMatchSnapshot();
    }
  });
});

describe("Submitting", () => {
  test("should update user details", async () => {
    if (wrapper) {
      window.URL.createObjectURL = jest.fn();
      const photoFile = new File([file], "test.png", { type: "image/png" });
      expect(wrapper.exists(".user-form")).toBeTruthy();

      // set values
      triggerInputChange(wrapper, { name: "firstName" }, "test");
      triggerInputChange(wrapper, { name: "lastName" }, "test");
      wrapper
        .find("#avatar")
        .hostNodes()
        .simulate("change", { target: { files: [photoFile] } });

      // submit form
      wrapper.find("form").simulate("focus").simulate("submit");
      expect(props.updateUserData).toHaveBeenLastCalledWith({
        displayName: "test test",
        photoURL: null,
        tenantId: user.uid,
      });
    }
  });
});

describe("Canceling", () => {
  test("should close the modal by clicking on react modal overlay", () => {
    if (wrapper) {
      expect(wrapper.exists(".user-form")).toBeTruthy();
      expect(wrapper.find(".ReactModal__Overlay").exists()).toBeTruthy();
      wrapper.find(".ReactModal__Overlay").simulate("click");
      expect(props.toggleUserModal).toHaveBeenLastCalledWith(false);
    }
  });
});
