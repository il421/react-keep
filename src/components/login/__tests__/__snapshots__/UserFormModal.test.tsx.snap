// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render LoginPage correctly 1`] = `
"<UserFormModal updateUserData={[Function: mockConstructor]} toggle={[Function: mockConstructor]} auth={{...}} modals={{...}}>
  <Modal isOpen={false} onRequestClose={[Function]} className=\\"user-modal\\" ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" role=\\"dialog\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} parentSelector={[Function: parentSelector]}>
    <Portal containerInfo={{...}}>
      <ModalPortal defaultStyles={{...}} isOpen={false} onRequestClose={[Function]} className=\\"user-modal\\" ariaHideApp={false} portalClassName=\\"ReactModalPortal\\" bodyOpenClassName=\\"ReactModal__Body--open\\" role=\\"dialog\\" closeTimeoutMS={0} shouldFocusAfterRender={true} shouldCloseOnEsc={true} shouldCloseOnOverlayClick={true} shouldReturnFocusAfterClose={true} parentSelector={[Function: parentSelector]} style={{...}} />
    </Portal>
  </Modal>
</UserFormModal>"
`;
